src/operator/numpy/linalg/np_gesvd-inl.h

```diff
@@ -0,0 +1,298 @@
+                                  DType* L, int ldl, DType* dL, int lddl,
+                                  DType* dA, int ldda, DType* V, int ldv) {
+    const int offx(k * m * ldx);
+    const int offl(k * ldl);
+    const int offdl(k * lddl);
+    const int offda(k * m * ldda);
+    const int offv(k * m * ldv);
+    const DType eps(gesvd_back_helper_eps(X));
+    DType denom1(0.0), denom2(0.0), elem(0.0);
+
+    for (int i = 0; i < m - 1; ++i) {
+      for (int j = i + 1; j < m; ++j) {
+        denom1 = L[offl + i] - L[offl + j];
+        denom2 = L[offl + i] + L[offl + j];
+        if (denom1 < eps) denom1 = eps;
+        if (denom2 < eps) denom2 = eps;
+        elem = (X[offx + i * ldx + j] - X[offx + j * ldx + i]) / denom1 / denom2;
+        X[offx + i * ldx + j] = elem * L[offl + j];
+        X[offx + j * ldx + i] = elem * L[offl + i];
+      }
+    }
+    for (int i = 0; i < m; ++i) {
+      elem = DType(0.0);
+      for (int j = 0; j < n; ++j) {
+        elem += dA[offda + i * ldda + j] * V[offv + i * ldv + j];
+      }
+      elem = -elem + dL[offdl + i];
+      X[offx + i * ldx + i] = elem;
+    }
+  }
+};
+
+struct gesvd_backward {
+  template<typename xpu, typename DType>
+  static void op(const Tensor<xpu, 3, DType>& dUT,
+                 const Tensor<xpu, 2, DType>& dL,
+                 const Tensor<xpu, 3, DType>& dV,
+                 const Tensor<xpu, 3, DType>& UT,
+                 const Tensor<xpu, 2, DType>& L,
+                 const Tensor<xpu, 3, DType>& V,
+                 const Tensor<xpu, 3, DType>& dA,
+                 const Tensor<xpu, 3, DType>& tempMs,
+                 const Tensor<xpu, 3, DType>& tempMr,
+                 Stream<xpu>* s, const nnvm::NodeAttrs& attrs) {
+    // Backward of (UT, L, V) = gesvd(A)
+    using namespace mxnet_op;
+    if (dA.dptr_ != dV.dptr_) {
+      Copy(dA, dV, s);
+    }
+    // From here on, we work on dA only
+
+    // Need temporal space, same shape as dUT
+    // invdV:
+    Kernel<GesvdBackHelper_dV, xpu>::Launch
+      (s, dA.size(0), dA.size(1), dA.size(2), L.dptr_, L.stride_, dA.dptr_, dA.stride_);
+
+    // G1:
+    // This copy is just to make sure there are no invalid values (NaN, infinity) in tempM
+    Copy(tempMs, dUT, s);
+    Copy(tempMr, dA, s);
+    gemm::op(dA, V, tempMs, DType(1.0), DType(0.0), false, true, s);
+    Kernel<GesvdBackHelper_G1, xpu>::Launch
+      (s, dA.size(0), dA.size(1), dA.size(2), tempMs.dptr_, tempMs.stride_,
+       L.dptr_, L.stride_);
+    gemm::op(dUT, UT, tempMs, DType(1.0), DType(1.0), true, false, s);
+
+    // G2:
+    Kernel<GesvdBackHelper_G2, xpu>::Launch
+      (s, dA.size(0), dA.size(1), dA.size(2), tempMs.dptr_, tempMs.stride_,
+       L.dptr_, L.stride_, dL.dptr_, dL.stride_, dA.dptr_, dA.stride_,
+       V.dptr_, V.stride_);
+
+    // G3:
+    gemm::op(tempMs, V, dA, DType(1.0), DType(1.0), false, false, s);
+    gemm::op(UT, dA, tempMr, DType(1.0), DType(0.0), false, false, s);
```

Comment:
It is very annoying we need this large temp space, because we don't have in-place left-multiply with square matrix. The drawback here is that for large matrices (large n), this needs a large temp space. It may be worth avoiding that.